service/shadowsocks.go

```diff
@@ -154,16 +135,31 @@ func WithPacketListener(listener transport.PacketListener) Option {
 
 // HandleStream handles a Shadowsocks stream-based connection.
 func (s *ssService) HandleStream(ctx context.Context, conn transport.StreamConn) {
-	var connMetrics TCPConnMetrics
+	var metrics TCPConnMetrics
 	if s.metrics != nil {
-		connMetrics = s.metrics.AddOpenTCPConnection(conn)
+		metrics = s.metrics.AddOpenTCPConnection(conn)
+	}
+	s.sh.Handle(ctx, conn, metrics)
+}
+
+// HandleAssociation handles a Shadowsocks packet-based connection.
+func (s *ssService) HandleAssociation(conn net.Conn) error {
+	assoc, err := s.NewPacketAssociation(conn)
+	if err != nil {
+		return fmt.Errorf("failed to handle association: %v", err)
 	}
-	s.sh.Handle(ctx, conn, connMetrics)
+	HandleAssociation(assoc)
```

Comment:
This sounds like you are using a default handler.
This should probably take an association and a handler.

I think the association is handling itself. That needs to be decoupled.